torch/onnx/_internal/fx/passes/modularization.py

```diff
@@ -0,0 +1,648 @@
+    ) -> bool:
+        """Determines if self is a superset of the provided module stack.
+
+        I.e., If self includes all elements from the provided module stack, plus additional
+        elements op top. If self is empty or root, this method always return False.
+        """
+        if self.is_empty_or_root():
+            return False
+
+        if module_stack.is_empty_or_root() is None:
+            return True
+
+        if len(self) <= len(module_stack):
+            return False
+
+        for i, parent_key in enumerate(module_stack):
+            if self[i] != parent_key:
+                return False
+
+        return True
+
+    def push(self, module_meta: _ModuleMeta) -> None:
+        """Pushes a module meta to the stack."""
+        self._module_stack.append(module_meta)
+
+    @_beartype.beartype
+    def is_same(
+        self,
+        module_stack: _ModuleStackMeta,
+    ) -> bool:
+        """Determines if self is the same as the provided module stack."""
+        return self._module_stack == module_stack._module_stack
+
+    @_beartype.beartype
+    def __eq__(self, __value: object) -> bool:
+        if not isinstance(__value, _ModuleStackMeta):
+            return False
+        return self.is_same(__value)
+
+    @property
+    def raw_meta(self) -> Optional[Dict[str, Tuple[str, type]]]:
+        """Returns the raw module stack meta data, i.e. node.meta['nn_module_stack']."""
+        return {
+            module_meta.raw_meta[0]: module_meta.raw_meta[1]
+            for module_meta in self._module_stack
+        }
+
+    def __repr__(self) -> str:
+        return f"ModuleStackMeta({self._module_stack})"
+
+    @property
+    def module_display_name(self) -> str:
+        """Returns the module display name of the top module."""
+        return self.top().module_display_name
+
+    @property
+    def qualified_module_class_name(self) -> str:
+        """Returns the qualified module class name of the top module."""
+        return self.top().qualified_module_class_name
+
+
+@functools.lru_cache()
+def _module_stack_meta_from_node(node: torch.fx.Node) -> _ModuleStackMeta:
+    return _ModuleStackMeta(node.meta.get("nn_module_stack"))
+
+
+def _get_unique_module_name(module_names: Dict[str, int], module_name: str) -> str:
+    module_names.setdefault(module_name, 0)
+    module_names[module_name] += 1
+    return f"{module_name}_{module_names[module_name]}"
+
+
+class _IRNode(abc.ABC):
+    """Base class for IR nodes.
+
+    IR nodes are used for modularization. They add a layer of abstraction on top of
+    torch.fx.Node.
+    """
+
+    @property
+    @abc.abstractmethod
+    def stack_meta(self) -> _ModuleStackMeta:
+        """The module stack meta data associated with this node."""
+        ...
+
+    @property
+    @abc.abstractmethod
+    def stack_trace(self) -> Optional[str]:
+        """The stack trace associated with this node."""
+        ...
+
+
+class _ModuleNode(_IRNode):
```

Comment:
Wait? Another IR (Sorry. I am strongly against new IR)? Is FX IR not sufficient for modularization?

Reply:
Added new [NOTE: Modularize Pass Implementation] section. 

The IRNode is not as intimidating as the name suggests. `_LeafNode` and `_ModuleNode` are just natural abstractions to help with this pass and make code look more responable, as compared to creating a bunch of nested list/tuple of fx.Nodes otherwise.